### PR TITLE
Tiered physical replicants

### PR DESCRIPTION
**Pull request by nirbhayc**
/plugin-branch tiered-physreps

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/sbuf2.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/reverse_conn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/phys_rep.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ plugins/reversesql/reversesql.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/reverse_conn.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
@@ plugins/reversesql/reversesql.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: <b>Error</b>. &#x26A0;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>6/464 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_timepart [core dumped]<br>sc_truncate<br>truncatesc<br>truncatesc_offline_generated<br>sc_addfield<br>op_makerecord<br>sc_timepart</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/sbuf2.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/reverse_conn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ plugins/reversesql/reversesql.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/reverse_conn.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
@@ plugins/reversesql/reversesql.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>6/468 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>logfill_logput_window_generated<br>logfill<br>truncatesc<br>truncatesc_offline_generated<br>ddl_analyze<br>tunables</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/sbuf2.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/reverse_conn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ plugins/reversesql/reversesql.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/reverse_conn.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
@@ plugins/reversesql/reversesql.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>4/469 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate_lockorder_generated<br>sc_truncate<br>selectv_deadlock<br>tunables</p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ cdb2api/cdb2api.c @@
+ Please port the change to cdb2jdbc if necessary.
+ If you add any internal configuration state to this file, please update the reset_the_configuration() function as well to include it.
@@ db/db_tunables.h @@
+ Please update the tunable documentation (docs/pages/config/config_files.md) accordingly.
@@ util/sbuf2.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
@@ db/reverse_conn.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ plugins/reversesql/reversesql.c @@
+ Please make sure that a mutex is unlocked before returning.
@@ db/reverse_conn.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
@@ plugins/reversesql/reversesql.c @@
+ Please replace any pthread_xxx() function in your diff with corresponding Pthread_xxx() wrapper.
```

File: bdb/rep.c [hunk 1]
```diff
@@ -427,7 +428,7 @@ void bdb_transfermaster(bdb_state_type *bdb_state)
     }
 }
 
-int gbl_set_coherent_state_trace = 0;
+int gbl_set_coherent_state_trace = 1;
```
Comment (nirbhayc) [inline bdb/rep.c:431]: Turn off

File: berkdb/rep/rep_record.c [hunk 2]
```diff
@@ -77,7 +77,9 @@ extern int gbl_reallyearly;
 extern int gbl_rep_process_txn_time;
 extern int gbl_is_physical_replicant;
 extern int gbl_dumptxn_at_commit;
-int gbl_rep_badgen_trace;
+extern int gbl_physrep_master_election_priority;
+
+int gbl_rep_badgen_trace = 1;
```
Comment (nirbhayc) [inline berkdb/rep/rep_record.c:82]: Turn off

File: berkdb/rep/rep_record.c [hunk 3]
```diff
@@ -843,6 +846,7 @@ __rep_enqueue_log(dbenv, rp, rec, gen)
 	DBT *rec;
 	uint32_t gen;
 {
+        //printf("==>%s:%d\n", __func__, __LINE__);
```
Comment (nirbhayc) [inline berkdb/rep/rep_record.c:849]: Remove

File: berkdb/rep/rep_record.c [hunk 4]
```diff
@@ -1091,6 +1095,8 @@ __rep_process_message(dbenv, control, rec, eidp, ret_lsnp, commit_gen, online)
 	if (LOG_SWAPPED())
 		__rep_control_swap(rp);
 
+        //logmsg(LOGMSG_USER, "%s:%d processing %d\n", __func__, __LINE__, rp->rectype);
```
Comment (nirbhayc) [inline berkdb/rep/rep_record.c:1098]: Remove

File: berkdb/rep/rep_record.c [hunk 5]
```diff
@@ -1356,6 +1365,9 @@ __rep_process_message(dbenv, control, rec, eidp, ret_lsnp, commit_gen, online)
 		}
 	}
 
+	//logmsg(LOGMSG_USER, "%s:%d processing %d\n", __func__, __LINE__,
+        //       rp->rectype);
```
Comment (nirbhayc) [inline berkdb/rep/rep_record.c:1369]: Remove

File: berkdb/rep/rep_record.c [hunk 6]
```diff
@@ -1560,6 +1572,9 @@ __rep_process_message(dbenv, control, rec, eidp, ret_lsnp, commit_gen, online)
 	case REP_LOG_MORE:
 		last_fill = comdb2_time_epochms();
 	case REP_LOG:
+		//logmsg(LOGMSG_USER, "%s:%d processing REPL_LOG (%d) message lsn: [%d][%d]\n",
+                //       __func__, __LINE__, rp->rectype,
```
Comment (nirbhayc) [inline berkdb/rep/rep_record.c:1576]: Remove

File: db/sqlinterfaces.c [hunk 7]
```diff
@@ -1320,7 +1320,8 @@ static void sql_statement_done(struct sql_thread *thd, struct reqlogger *logger,
         char zFingerprint2[FINGERPRINTSZ*2+1];
         memset(zFingerprint1, 0, sizeof(zFingerprint1));
         memset(zFingerprint2, 0, sizeof(zFingerprint2));
-        if ((memcmp(clnt->work.aFingerprint, zFingerprint1, FINGERPRINTSZ) != 0) &&
+        // TODO (NC): Temporary hack to suppress this annoying warning, fixme.
```
Comment (nirbhayc) [inline db/sqlinterfaces.c:1323]: Remove

File: plugins/reversesql/reversesql.c [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+    1. P opens a connection to D.
+    2. P sends "reversesql" request over this connection and ..
+        2.1. Waits for "newsql" request to show up back on the connection
+    3. D receives and processes the "reversesql" request by ..
+        3.1. Passing the connection's fd to cdb2_open(fd) and
+        3.2. Executing queries via the cdb2 handle initialized above.
+
+    TODOs:
+       * Only allow readonly queries
```
Comment (nirbhayc) [inline plugins/reversesql/reversesql.c:35]: REmove

File: db/phys_rep.c [hunk 9]
```diff
@@ -24,240 +43,374 @@ typedef struct DB_Connection {
...
 
-        if (do_truncate) {
-            info = get_last_lsn(thedb->bdb_env);
-            prev_info = handle_truncation(repl_db, info);
-            if (prev_info.file == 0) {
-                close_repl_connection();
+static int get_local_hndl(cdb2_hndl_tp **hndl) {
+    int rc = cdb2_open(hndl, gbl_dbname, "local", 0);
+    if (rc != 0) {
+        physrep_logmsg(LOGMSG_ERROR, "%s:%d Failed to connect to %s@%s (rc: %d)\n",
```
Comment (nirbhayc) [inline db/phys_rep.c:295]: Cleanup handle.

File: db/reverse_conn.c [hunk 10]
```diff
@@ -0,0 +1,626 @@
...
+                __func__, new_fd, errno, strerror(errno));
+        rc = -1;
+        goto cleanup;
+    }
+#endif
+
+    sbuf2setbufsize(sb, netinfo_ptr->bufsz);
+
+    char msg[120] = {0};
+    snprintf(msg, sizeof(msg), "reversesql\n%s\n%s\n%s\n",
```
Comment (nirbhayc) [inline db/reverse_conn.c:179]: Use new protobuf hello message.

**Review by markhannum — DISMISSED** [review]
Reviewed extensively in person over several days .. this is good to merge .. !

**Review by markhannum — APPROVED** [review]

**Pull request merged**